frontend/src/plugins/api.js

```diff
@@ -104,15 +104,26 @@ class API {
         }
 
         try {
+            const controller = new AbortController();
+            const timeout = setTimeout(() => controller.abort(), timeout_s);
+
             const response = await fetch(url, {
                 method: method || 'GET',
                 body: encoded_body,
-                headers: allheaders
+                headers: allheaders,
+                signal: controller.signal,
             });
+
+            clearTimeout(timeout);
             await this._is_response_ok(response);
             return await response.json();
         } catch (e) {
-            this._error_to_alert(e);
+            if(e.name === "AbortError") {
+                this._error_to_alert({message: "Request timed out"});
+                console.error("request timed out: %s", url);
+            } else {
+                this._error_to_alert(e);
+            }
```

Comment:
Thank you very much for your contribution and my apologies for not picking this up earlier.

In principle, the solution looks perfectly fine. I'd only like to ask you small favor of you: Would you mind moving the handling of `AbortError` into the `_error_to_alert` method?

Reply:
i was considering this approach but it had some drawbacks:
- the `AbortError` that this code produces is quite _ugly_ and does not show that there was a request timeout. this is a data point that the user should be notified about, instead of expecting them to read a long message, or even make sense of the word `AbortError`.
- considering this issue, if i display a message saying that a request timed out in the `_error_to_alert` function, it could be misleading if the code were to throw an `AbortError` for another reason.

so i'd request some feedback on this before i proceed to implement it.

Reply:
Yes, those are very good arguments.

I had a quick look at the docs of [AbortController](https://developer.mozilla.org/en-US/docs/Web/API/AbortController/abort), which gave me another idea:

How about, we leave the content of the `catch` block unchanged and instead use your self-defined object as argument to the call to the `abort()` method in line 108?

E.g.:
```js
            const timeout = setTimeout(() => controller.abort(
              {"message": "Request timed out."}), 
              timeout_s
            );

Reply:
apparently, the `controller.abort(reason)` doesn't make it throw the reason object or string. it just sets the value of `controller.signal.reason` instead.

for example, the following code:
```js
async function delay (ms, value) {
  return new Promise(res => setTimeout(() => res(value), ms));
}

async function example () {
  const controller = new AbortController();
  try{
		const timeout = setTimeout(() => controller.abort(
      {"message": "Request timed out"}), 1);
    const response = await fetch("https://google.com", {signal: controller.signal});
		await delay(2000);
    console.log(controller.signal.reason);
	} catch (e) {  
    console.log(e.constructor);
    console.log(controller);
    console.log(controller.signal);
	}
}

console.log("calling method");
example();
```

produces this output in the console:
```
"calling method"
function DOMException() { [native code] }
[object AbortController] {
  abort: function abort() { [native code] },
  signal: [object AbortSignal] {
    aborted: true,
    addEventListener: function addEventListener() { [native code] },
    dispatchEvent: function dispatchEvent() { [native code] },
    onabort: null,
    reason: { ... },
    removeEventListener: function removeEventListener() { [native code] },
    throwIfAborted: function throwIfAborted() { [native code] }
  }
}
[object AbortSignal] {
  aborted: true,
  addEventListener: function addEventListener() { [native code] },
  dispatchEvent: function dispatchEvent() { [native code] },
  onabort: null,
  reason: {
    message: "Request timed out"
  },
  removeEventListener: function removeEventListener() { [native code] },
  throwIfAborted: function throwIfAborted() { [native code] }
}
```

`throwIfAborted` is interesting, but it will require a lot more tinkering.

Reply:
To be honest, I had not bothered to double check the implementation of `AbortController`, but when I tested it in Firefox, the catched object `e` was identical to the object passed into `abort()`.

In my case, this code:

```js
        const reason = {message: "Request timed out"};
        try {
            const controller = new AbortController();
            const timeout = setTimeout(() => controller.abort(reason), timeout_s);
            // ...
        } catch (e) {
            console.log("Time out", e, typeof e, e === reason)
            this._error_to_alert(e);
            return null;
        }
```
produces this output on the JS console of Firefox:
![grafik](https://github.com/SUSE/aimaas/assets/1128117/282f5dc6-aa23-4038-9b1f-57157399d778)

---

And now for the fun part: Chrome behaves differently, here `e` is a `DOMException`, which does not contain teh reason. :face_with_head_bandage: 

With that in mind, my proposal really is not alternative.

Reply:
it's interesting that it ran on Firefox lol.

Reply:
Indeed, one might even wonder whether this is the beginning of another mess similar to Internet Explorer :rofl: